reports/storing-large-real-time-data-streams-in-pod-using-LDES.md

```diff
@@ -0,0 +1,207 @@
+## Problem
+<!--
+You can reuse the pitch of the challenge, but check if you need to make changes.
+For example, it might happen that the approved solution does more than what the original pitch requested.
+-->
+
+Data streams are becoming omnipresent and 
+are a crucial component in many use cases. 
+Storing streams in a low-cost file-based Web environment could be done using 
+[Linked Data Event Streams](https://semiceu.github.io/LinkedDataEventStreams/) (LDES). 
+However, pushing large volumes of high volatile data into a Solid-based LDES is not possible because 
+the [current solution](https://github.com/woutslabbinck/SolidEventSourcing) does the partitioning of the data 
+after all data has been retrieved, 
+instead of in a streaming fashion. 
+This crashes the [Community Solid Server](https://github.com/CommunitySolidServer/CommunitySolidServer) 
+due to the high load on the server 
+when repartitioning large amounts of data.
+We use data from the [DAHCC dataset](https://dahcc.idlab.ugent.be/) for this challenge,
+which contains data streams describing the behaviour of various patients.
+The streams contain over 100.000 events.
+
+We want a streaming Solid-based LDES connector that replays data and 
+partitions this data in a streaming fashion when retrieving the data, 
+instead of needing to wait till the whole dataset is received.
+Besides avoid the high load on the server,
+replaying allows to mimic the real-time behaviour of data streams,
+even though the data is historical data.
+This allows to showcase how solutions can process live data streams and
+how they can handle different data rates.
+
+## Approved solution
+<!--
+Provide information about the approved solution:
+names of tools/libraries created, repos, and so on.
+-->
+
+We developed a [tool](https://github.com/SolidLabResearch/LDES-in-SOLID-Semantic-Observations-Replay/) that
+replays captured streams. 
+The tool has the following features:
+
+- The tool is pull-based as it pulls the observations from n-triples files.
+- The tool adds new events to the right bucket.
+- If new buckets are required, the tool updates the LDES.
+- The user can parameterize the bucket size.
+
+The repository contains both a [Web app](https://github.com/SolidLabResearch/LDES-in-SOLID-Semantic-Observations-Replay/tree/main/webapp) and
+a [engine](https://github.com/SolidLabResearch/LDES-in-SOLID-Semantic-Observations-Replay/tree/main/engine), but
+only the engine is relevant for [challenge #82](https://github.com/SolidLabResearch/Challenges/issues/82) and
+this report.
+
+The solution builds on [this code](https://github.com/woutslabbinck/SolidEventSourcing).
+The code is not included as a dependency as it was not designed as a library.
+We copy and pasted the parts of the code that were useful for our solution.
+
+<!--
+Provide a list of important technical decisions and assumptions.
+-->
+
+We made the following important technological decisions and assumptions:
+- TODO: @Stijn Can you add this?
```

Comment:
@svrstich Can you also provide me with this?

Reply:
Of course, sorry for the delay :-s

- JavaScript / TypeScript as main development language to be in line with current practices.
- LDES as a means to represent a stream of events detected by IOT-related devices.
- Paginated implementation to support loaded large datasets.
- n-triples files are datasource, as they represent the state of the devices and observations as originally persisted using Apache Feather.
- Single-user demonstration implementation, as the main goal is validating the approach, not developing a industry-ready implementation.
- https://www.npmjs.com/package/n3 for streaming, high-throughput handling of RDF data.
- an recursive implementation of the mergeSort algorithm to sort the observation based on timestamp (configurable).
- works with a pointer-based algorithm, which can be advanced according to the ordering by the end user/automatically.
- includes an optimisation to manage the size of the pods.